GHC/Cmm/Dominators.hs

```diff
@@ -0,0 +1,170 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+
+module GHC.Cmm.Dominators
+  ( DominatorSet(..)
+  , GraphWithDominators(..)
+  , RPNum
+  , graphWithDominators
+
+  , graphMap
+  , gwdRPNumber
+  , gwdDominatorsOf
+  , dominatorsMember
+
+  , intersectDominators
+  )
+where
+
+import Data.Array.IArray
+import Data.Foldable()
+
+import qualified Data.IntMap.Strict as IM
+import qualified Data.IntSet as IS
+
+import qualified GHC.CmmToAsm.CFG.Dominators as LT
+
+import GHC.Cmm.Dataflow
+import GHC.Cmm.Dataflow.Block
+import GHC.Cmm.Dataflow.Collections
+import GHC.Cmm.Dataflow.Graph
+import GHC.Cmm.Dataflow.Label
+import GHC.Cmm
+
+import GHC.Utils.Outputable(Outputable(..), text, int, hcat, (<+>))
+import GHC.Utils.Panic
+
+
+-- | =Dominator sets
+--
+-- Node X dominates node Y if and only if every path from the entry to
+-- Y includes X.  Node Y technically dominates itself, but it is
+-- never included in the *representation* of its dominator set.
+--
+-- A dominator set is represented as a linked list in which each node
+-- points to its *immediate* dominator, which is its parent in the
+-- dominator tree.
+
+data DominatorSet = ImmediateDominator { ds_label  :: Label
+                                       , ds_parent :: DominatorSet
+                                       }
```

Comment:
if we add a `UniqSet Label` here to cache all dominators, we can implement O(1) `dominatorsMember`. whether that's a worthy optimization or not, i don't have very strong opinion, and let's hear what ghc people think later.

Reply:
Most of the time you want the immediate dominator, which is there in constant time.  And that's also what the algorithm produces.  So we could certainly provide a function of type `DominatorSet -> LabelSet`, but I'm inclined not to unless someone asks.